### PR TITLE
Enable user defined extra-constraints to inject into minizinc

### PR DESCRIPTION
**Pull request by philzook58**
This puts an extra field into the patch config "extra-constraints"which will inject verbatim the string given to the tail of the minizinc model on a per patch basis.
This feature might supersede the excluded registers features and preassignment

~~The huge caveat is that currently minizinc receives non user friendly names. This can be somewhat ameliorated by using `forall` sometimes. You also need to have knowledge of how the model works to use this feature. Both require expert use. Only for the brave and true atm. We could work on maintaining good names and a library of helper predicates if this is a desirable feature.~~

I have removed most vestiges of s-expressions from the minizinc serialization. It was convenient but also horrible to debug.
I added a new parameter to mzn_params `hvars_temps` which maps from high level variables names to temporaries that make up this high level variable name.
With this, there are user friendly names available for building extra-constraints

As a side note, i had to move some data types to Data.ml to resolve a dependency ordering issue and removed exception throwing from `Linear_ssa.orig_name` instead using an `option`

### CODE REVIEW

File: bap-vibes/src/minizinc.ml [hunk 1]
```diff
@@ -240,8 +227,12 @@ let serialize_mzn_params
             | Some s -> Set.add s t1)) in
   let temps = params.temps |> Var.Set.to_list in
   let temp_names =
-    List.map ~f:(fun t -> Var.sexp_of_t t |> Sexp.to_string) temps
+    List.map ~f:(fun t -> Var.to_string t) temps
   in
+  let hvars = List.map temps ~f:(fun temp -> Linear_ssa.orig_name (Var.to_string temp)) in
+  let hvars = String.Set.of_list (List.filter_opt hvars) in
+  let hvars = String.Set.diff hvars (String.Set.of_list (gpr @ regs)) in
```
Comment (philzook58) [inline bap-vibes/src/minizinc.ml:234]: Hmm. Maybe I don't need to do this now that the convention is referring to "hvar_" prepended variables. I don't know that you'd want to force `hvar_SP` to be anything though for example

File: bap-vibes/src/minizinc.ml [hunk 2]
```diff
@@ -240,8 +227,12 @@ let serialize_mzn_params
             | Some s -> Set.add s t1)) in
   let temps = params.temps |> Var.Set.to_list in
   let temp_names =
-    List.map ~f:(fun t -> Var.sexp_of_t t |> Sexp.to_string) temps
+    List.map ~f:(fun t -> Var.to_string t) temps
   in
+  let hvars = List.map temps ~f:(fun temp -> Linear_ssa.orig_name (Var.to_string temp)) in
```
Comment (bmourad01) [inline bap-vibes/src/minizinc.ml:232]: This could be a `List.filter_map`

File: bap-vibes/tests/integration/test_minizinc.ml [hunk 3]
```diff
@@ -56,9 +56,9 @@ let (:=) lhs opcode = fun args -> simple_op' opcode lhs args
 let mov = Ir.Opcode.create ~arch:"arm" "mov"
 
 (* Define some temporaries to use in a test. *)
-let t1 = Var.create "t1" (Imm 32)
-let t2 = Var.create "t2" (Imm 32)
-let t3 = Var.create "t3" (Imm 32)
+let t1 = Var.create "_000004ff_t1" (Imm 32)
```
Comment (bmourad01) [inline bap-vibes/tests/integration/test_minizinc.ml:59]: I'm wondering how fragile these TID-prefixed names are. Can we trust that they will be consistent across different runs?

Comment (philzook58) [reply bap-vibes/tests/integration/test_minizinc.ml:59]: I don't think they are fragile. I just needed orig_name to return Some instead of none

**Pull request merged**